### DISCUSSION TITLE
[ZK Stack] Server suddenly stops processing new batches

### DISCUSSION THREAD

**Discussion by therealdannzor** [zkSync-Community-Hub/zksync-developers#472] [ZK Stack] [answered]
### Module Affected

- [ ] System contracts (era-system-contracts)
- [ ] Smart contracts (era-contracts)
- [ ] zkEVM (era-zk_evm)
- [X] Server (zksync-era)
- [ ] Prover (era-boojum)
- [ ] Fee withdrawer (era-fee-withdrawer)
- [ ] Block Explorer (clock-explorer)
- [ ] Portal - Wallet + Bridge (dapp-portal)
- [ ] Other (please specify below)

### Rust Version

1.75.0

### Operating System

- [ ] Ubuntu
- [ ] MacOS
- [ ] Windows
- [ ] Other (Please specify below)

### Issue Description

We have a public zkSync Era hyperchain test network with version `core-v20.2.0`.

Regrettably, our network is experiencing a critical issue during normal operations. Specifically, when bridging ETH from L1->L2 and moving back from L2->L1, the network halts its batch execution, significantly impacting its functionality.

The deployed [dapp-portal](https://github.com/matter-labs/dapp-portal) is the only external component to bridge Sepolia and our test network.

This can be observed in the following two transactions:
1. [L1 (Sepolia)](https://sepolia.etherscan.io/tx/0x9603016bad5f38d21b2f6108c3071249fa035d4ec62be693ac8e103eedb8a9fd)
2. [L2 (PlayFi)](https://albireo-explorer.playfi.ai/tx/0xd8518937a0a3deb5f312fbbccd011b1acd4e8dae4833dfaa32b5176b92cff5b1#overview)

As you can see, the L2 transaction is still displayed as sending on Ethereum despite being over three days old. We have not modified the underlying code in any way to potentially compromise the node's liveness.

### Expected Behavior

Batches are processed as expected as new L2 blocks are formed

### Current Behavior

No new batches are being processed despite transactions taking place

### Additional Details

```
12 April 2024 at 02:02 (UTC+2:00)	[2m2024-04-12T00:02:36.737588Z[0m [32m INFO[0m [1mloop_iteration[0m[1m{[0m[3mprevious_block[0m[2m=[0mL1BlockNumber(5678193)[1m}[0m[2m:[0m [2mzksync_core::eth_sender::eth_tx_manager[0m[2m:[0m eth_tx 1994 with hash 0x4e555aad076f2aa026debf7a491c503c146787e19ae48a9c8f54183f8c76cc3f for ExecuteBlocks is confirmed. Gas spent: 81950	a41c866820b845fbaf6227306c17e277	albireo-zksync-core
12 April 2024 at 02:02 (UTC+2:00)	[2m2024-04-12T00:02:21.677397Z[0m [32m INFO[0m [1mloop_iteration[0m[2m:[0m [2mzksync_core::eth_sender::eth_tx_aggregator[0m[2m:[0m eth_tx with ID 1994 for op execute was saved for L1 batches L1BatchNumber(705)..=L1BatchNumber(705)	a41c866820b845fbaf6227306c17e277	albireo-zksync-core
12 April 2024 at 02:02 (UTC+2:00)	[2m2024-04-12T00:02:01.551218Z[0m [32m INFO[0m [1mloop_iteration[0m[1m{[0m[3mprevious_block[0m[2m=[0mL1BlockNumber(5678190)[1m}[0m[2m:[0m [2mzksync_core::eth_sender::eth_tx_manager[0m[2m:[0m eth_tx 1993 with hash 0xf7235427f278acca4dfd653dd2e81186fc316a2fdbb567bf4b8e5307533396cd for PublishProofBlocksOnchain is confirmed. Gas spent: 68065	a41c866820b845fbaf6227306c17e277	albireo-zksync-core
12 April 2024 at 02:02 (UTC+2:00)	[2m2024-04-12T00:02:01.517060Z[0m [32m INFO[0m [1mloop_iteration[0m[1m{[0m[3mprevious_block[0m[2m=[0mL1BlockNumber(5678190)[1m}[0m[2m:[0m [2mzksync_core::eth_sender::eth_tx_manager[0m[2m:[0m eth_tx 1992 with hash 0x5ecd3bf1d878c719d0c19840b56fc13f082b9ed31433523eae3fdf2217d882d2 for CommitBlocks is confirmed. Gas spent: 111887	a41c866820b845fbaf6227306c17e277	albireo-zksync-core
12 April 2024 at 02:01 (UTC+2:00)	[2m2024-04-12T00:01:40.088692Z[0m [32m INFO[0m [1mloop_iteration[0m[2m:[0m [2mzksync_core::eth_sender::eth_tx_aggregator[0m[2m:[0m eth_tx with ID 1993 for op proof was saved for L1 batches L1BatchNumber(705)..=L1BatchNumber(705)	a41c866820b845fbaf6227306c17e277	albireo-zksync-core
12 April 2024 at 02:01 (UTC+2:00)	[2m2024-04-12T00:01:34.762576Z[0m [32m INFO[0m [1mloop_iteration[0m[2m:[0m [2mzksync_core::eth_sender::eth_tx_aggregator[0m[2m:[0m eth_tx with ID 1992 for op commit was saved for L1 batches L1BatchNumber(705)..=L1BatchNumber(705)
```

The last processed tx has id `1994` and is included in batch `705`. After this, there are no more logs concerning the next batch `706` which is still pending, as seen in the link above on the L2 side.

### Prior Research

- [X] Yes, I searched for similar issues and did not find any matching ones.

**Comment by bxpana** [comment]
Thanks for reporting. Will check on this for you

**Comment by Archethect** [comment]
@bxpana 
While doing some research we noticed that `eth_tx_aggregator` stopped... although we don't see any signals hinting us about stopping that service in the logs.

**Reply by bxpana** [reply]
Good to know. Will let the team know about this part too.

**Reply by Archethect** [reply]
Trying to kickstart the server now, but it seems too much time has passed and the prepared L1 transactions to commit, prove and execute are now failing because the delay in execution. 

-->
`/// @dev Timestamp - seconds since unix epoch.
uint256 constant COMMIT_TIMESTAMP_NOT_OLDER = 3 days;`

Is there a way to kickstart the process again since the last succesful batch and pack the pending tx in new batches?

I'm sure we can modify some DB records to make it work again, but I have no idea which ones precisely.

**Reply by bxpana** [reply]
@Archethect @therealdannzor Checked with the team and we’re currently unable to provide an immediate fix, but we’re working on releasing a new zk tooling version that should help. We’ll keep you updated on the progress!

Thanks for your support and patience

**Reply by bxpana** [reply]
We'll have an update about the zk tooling either here or on https://twitter.com/zkSyncDevs

**Reply by therealdannzor** [reply]
> We'll have an update about the zk tooling either here or on https://twitter.com/zkSyncDevs

Thank you for your response and update.

Look forward to trying out the updated tooling. I assume this concerns the `zk stack` command 🙂.

**Reply by Archethect** [reply]
Hey @bxpana ,
Thanks for the info. However, its kind of critical for us because our chain is halted atm (although it is a testnet... Its just not a good look.) I appreciate the fact that a tool will be developed, but is there at least a way to get it up and running now again? I don't care to fiddle with the database or to do some other stuff... A very simple 'how to' would suffice. It doesnt need to be very detailed wnd I'm willing to put some research in it myself.

**Reply by bxpana** [reply]
@Archethect I'll relay that back to the team and see what more I can find out.

**Reply by RomanBrodetski** [reply]
Hello @Archethect first of all - I'm sorry that you are experiencing this issue, and indeed we are working on a new way to init and run hyperchains which is designed with robustness in mind.

Now, re: current issue - I think you have the maximum number of transacation "in flight" or "unconfrimed" on l1. And indeed, you'll need to look into the DB to figure out what exactly is going on.

Please do `select count(*), confirmed_eth_tx_history_id IS NULL from eth_txs group by confirmed_eth_tx_history_id IS NULL` and paste the result here - this will show how many confirmed - and unconfirmed transcations you have.

If you indeed have many (I thnk default limit is 30), you should figure out why are they not being confirmed. For this find one transaction for which confirmed_eth_tx_history_id IS NULL, find the correpsonding entries in `eth_txs_history` table and try to find their hashes on etherscan.

**Reply by l3erryy99** [reply]
This run test net right?

**Reply by Archethect** [reply]
Hello @RomanBrodetski  and thanks for following up.

We indeed have 30 pending tx in our DB. See results below:

"count"	"?column?"
"1994"	"false"
"30"	"true"

However, I think this filled up because our first tx that got stuck (tx 1995) is not going through on Sepolia. From what I can understand, this tx (which you can see here: https://sepolia.etherscan.io/tx/0x9c1e8b76d47ffb8e1938cb2d71eeda297e1fa67616886688dd2f9a81a1d60bc6) failed because too much time has passed between the previously successful batch and the new batch that needs to be committed.

See the code snippet below for the code it tries to execute:

```
// All L2 blocks have timestamps within the range of [batchTimestamp, lastL2BlockTimestamp].
// So here we need to only double check that:
     // - The timestamp of the batch is not too small.
     // - The timestamp of the last L2 block is not too big.
 
// New batch timestamp is too small
 require(block.timestamp - COMMIT_TIMESTAMP_NOT_OLDER <= batchTimestamp, "h1"); 
 // The last L2 block timestamp is too big
require(lastL2BlockTimestamp <= block.timestamp + COMMIT_TIMESTAMP_APPROXIMATION_DELTA, "h2"); 
```

The reason is that for whatever reason, eth_tx_aggregator was down for several days before we restarted the chain and it kicked back in action, leaving a gap too big between batches for the commit on L1 to succeed.
        
So I think the question is... How can we solve this issue? I already tried with changing the timestamps of the batches in the DB manually, but this fails because the timestamp is also hashed inside the batches upfront and thus the hashes do not check out anymore.

**Reply by Archethect** [reply]
Any update on this one @RomanBrodetski ?

**Reply by RomanBrodetski** [reply, marked as answer]
yes, now it's impossible to submit this batch because too much time has passed. 

Changing timestamps will not help - you need to either start the chain over or rollback the problematic batch along with all the batches thereafter. We have a tool that helps with that: https://github.com/matter-labs/zksync-era/blob/main/core/bin/block_reverter/src/main.rs. It lacks documentation (and we are working on that), but you can infer the usage from CLI arguments. You don't need to rollback the smart contact state, only the local DB. 

Given that you need to rollback blocks anyway, it might be a simpler idea to just start the chain over - and make sure the delay of submitting batches to l1 doesn't exceed three days.

**Reply by Archethect** [reply]
Allright, thanks! 
That's what I needed. I'll try to revert using the tool.
Will keep you posted.

**Reply by Archethect** [reply]
> you need to either start the chain over or rollback the problematic batch along with all the batches thereafter. 

I was able to roll back the batches with the tool but rolling back the batches like this does not work because the timestamp of any successful batch right now is older than 3 days, right? 
So I think I'm just stuck at the last successful batch at the moment... (Because the delta between this successful batch and a new batch will always be > 3 days)
I'm trying to avoid restarting the chain as I try to avoid losing the state of the chain (although its a testnet).

I'm just worried that on a production system if for whatever reason the batching halts for > 3 days, that the chain is broken forever.

**Reply by Archethect** [reply]
@RomanBrodetski 

My bad, it seems to work now. I forgot to clear failed transactions in the DB after running the revert script. 
All good now. We are running again. 

Thanks so much for your support. I really appreciate it! 🙏

https://sepolia.etherscan.io/address/0x6a3de60625d3a2fd68cfb6a184a6fdcf90db7361

**Reply by RomanBrodetski** [reply]
oh, very happy that it works!

**Comment by therealdannzor** [comment]
Thanks @RomanBrodetski for your help in resolving this issue. cc @bxpana

Considering there's a way to do this [rollback the problematic batch], I assume this is not the first time happening. Do you know how this faulty behavior can be replicated? What I can remember, the only thing I was doing was bridging to/from L1 and L2, at times rapidly.

**Reply by RomanBrodetski** [reply]
maybe the block was mined, and then the node was turned off/crashed - before this block was submitted to L1. And then it was only started again in three days. Is it possible?

**Reply by Archethect** [reply]
From what I remember the server kept running, but eth_sender (eth_tx_aggregator) was not running anymore. We did not see anything related to it in the logs that could have caused it (just that there were no logs related to it anymore). However, it was certainly not a failed L1 tx because this just kills the complete server process. We noticed it too late and once we did, 3 days had passed.

**Reply by bxpana** [reply]
Glad @RomanBrodetski was able to help you guys. If you see this type of behavior again let us know and we can try to further debug.

**Comment by qiwenzhou** [comment]
I encountered  similar problem. Could you please share the commands that you use to revert blocks?
